prefect_gcp/aiplatform.py

```diff
@@ -0,0 +1,358 @@
+from google.cloud.aiplatform.gapic import JobServiceClient
+from google.cloud.aiplatform_v1.types.custom_job import (
+    ContainerSpec,
+    CustomJob,
+    CustomJobSpec,
+    Scheduling,
+    WorkerPoolSpec,
+)
+from google.cloud.aiplatform_v1.types.job_service import CancelCustomJobRequest
+from google.cloud.aiplatform_v1.types.job_state import JobState
+from google.cloud.aiplatform_v1.types.machine_resources import MachineSpec
+from google.protobuf.duration_pb2 import Duration
+from prefect.exceptions import InfrastructureNotFound
+from prefect.infrastructure import Infrastructure, InfrastructureResult
+from prefect.utilities.asyncutils import run_sync_in_worker_thread, sync_compatible
+from pydantic import Field
+from slugify import slugify
+from typing_extensions import Literal
+
+from prefect_gcp import GcpCredentials
+
+
+class VertexAICustomTrainingJobResult(InfrastructureResult):
+    """Result from a Vertex AI custom training job."""
+
+
+class VertexAICustomTrainingJob(Infrastructure):
+    """
+    Infrastructure block used to run Vertex AI custom training jobs.
+    """
+
+    _block_type_name = "Vertex AI Custom Training Job"
+    _block_type_slug = "vertex-ai-custom-training-job"
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/4CD4wwbiIKPkZDt4U3TEuW/c112fe85653da054b6d5334ef662bec4/gcp.png?h=250"  # noqa
+
+    type: Literal["vertex-ai-custom-training-job"] = Field(
+        "vertex-ai-custom-training-job", description="The slug for this task type."
+    )
+
+    gcp_credentials: GcpCredentials = Field(
+        default_factory=GcpCredentials,
+        description=(
+            "GCP credentials to use when running the configured "
+            "Vertex AI training job."
+        ),
+    )
+    region: str = Field(
+        default=...,
+        description="The region where the Vertex AI custom training job resides.",
+    )
+    image: str = Field(
+        default=...,
+        title="Image Name",
+        description=(
+            "The image to use for a new Vertex AI custom training job. This value must "
+            "refer to an image within either Google Container Registry "
+            "or Google Artifact Registry, like `gcr.io/<project_name>/<repo>/`"
+        ),
+    )
+    env: Dict[str, str] = Field(
+        default_factory=dict,
+        title="Environment Variables",
+        description="Environment variables to be passed to your Cloud Run Job.",
+    )
+    machine_type: str = Field(
+        default="n1-standard-4",
+        description="The machine type to use for the run, which controls the available "
+        "CPU and memory.",
+    )
+    accelerator_type: Optional[str] = Field(
+        default=None, description="The type of accelerator to attach to the machine."
+    )
+    maximum_run_time: datetime.timedelta = Field(
+        default=datetime.timedelta(days=7), description="The maximum job running time."
+    )
+    network: Optional[str] = Field(
```

Comment:
do you think we should allow configuration of some `reserved_IP_range` that would [restrict where the job gets deployed](https://cloud.google.com/vertex-ai/docs/reference/rest/v1/CustomJobSpec) within the provided `network`?